applications/FluidDynamicsApplication/custom_elements/d_vms_dem_coupled.cpp

```diff
@@ -101,6 +102,114 @@ void DVMSDEMCoupled<TElementData>::PrintInfo(std::ostream& rOStream) const
+    const typename TElementData::ShapeFunctionsType& rN) const
+{
+    return DVMS<TElementData>::GetAtCoordinate(rValues, rN);
+}
+
+template <class TElementData>
+double DVMSDEMCoupled<TElementData>::GetAtCoordinate(
+    const double Value,
+    const typename TElementData::ShapeFunctionsType& rN) const
+{
+    return DVMS<TElementData>::GetAtCoordinate(Value, rN);
+}
+
+template< class TElementData >
+void DVMSDEMCoupled<TElementData>::AlgebraicMomentumResidual(
+    const TElementData& rData,
+    const array_1d<double,3> &rConvectionVelocity,
+    array_1d<double,3>& rResidual)
+{
+    const GeometryType rGeom = this->GetGeometry();
+
+    Vector convection; // u * grad(N)
+    this->ConvectionOperator(convection,rConvectionVelocity,rData.DN_DX);
+
+    const double density = this->GetAtCoordinate(rData.Density,rData.N);
+    const double viscosity = this->GetAtCoordinate(rData.DynamicViscosity, rData.N);
+    Matrix permeability = this->GetAtCoordinate(rData.Permeability, rData.N);
+    Matrix inv_permeability = ZeroMatrix(Dim, Dim);
+    const auto& r_body_forces = rData.BodyForce;
+    const auto& r_velocities = rData.Velocity;
+    const auto& r_pressures = rData.Pressure;
+
+    double det_permeability = MathUtils<double>::Det(permeability);
+    MathUtils<double>::InvertMatrix(permeability, inv_permeability, det_permeability, -1.0);
+
+    Matrix sigma = viscosity * inv_permeability;
+
+    for (unsigned int i = 0; i < NumNodes; i++) {
+        const array_1d<double,3>& r_acceleration = rGeom[i].FastGetSolutionStepValue(ACCELERATION);
+        Vector sigma_U = ZeroVector(Dim);
+        for (unsigned int d = 0; d < Dim; d++) {
+            for (unsigned int e = 0; e < Dim; e++){
+                sigma_U[d] += sigma(d,e) * rData.N[i] * r_velocities(i,e);
+            }
+            rResidual[d] += density * ( rData.N[i]*(r_body_forces(i,d) - r_acceleration[d]) - convection[i]*r_velocities(i,d)) - rData.DN_DX(i,d)*r_pressures[i] - sigma_U[d];
+        }
+    }
+}
+
+template< class TElementData >
+void DVMSDEMCoupled<TElementData>::MomentumProjTerm(
+    const TElementData& rData,
+    const array_1d<double,3>& rConvectionVelocity,
+    array_1d<double,3> &rMomentumRHS)
+{
+    Vector AGradN;
+    this->ConvectionOperator(AGradN,rConvectionVelocity,rData.DN_DX);
+
+    const double density = this->GetAtCoordinate(rData.Density,rData.N);
+    const double viscosity = this->GetAtCoordinate(rData.DynamicViscosity, rData.N);
+    Matrix permeability = this->GetAtCoordinate(rData.Permeability, rData.N);
+    Matrix inv_permeability = ZeroMatrix(Dim, Dim);
+
+    double det_permeability = MathUtils<double>::Det(permeability);
+    MathUtils<double>::InvertMatrix(permeability, inv_permeability, det_permeability, -1.0);
+
+    Matrix sigma = viscosity * inv_permeability;
+
+    for (unsigned int i = 0; i < NumNodes; i++) {
+        for (unsigned int d = 0; d < Dim; d++) {
+            Vector sigma_u = ZeroVector(Dim);
```

Comment:
I think you can use an `array_1d<double,Dim>` in here. I'd also allocate the vector outside the loop. Then you can re-initialize it for each `d` iterator value.